### PR TITLE
EAD: Elastic-Net Attacks

### PR DESCRIPTION
**Pull request by araujoalexandre**
Add EAD: Elastic-Net Attacks from this paper : https://arxiv.org/abs/1709.04114

### CODE REVIEW

File: advertorch/attacks/ead.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        final_dist = torch.FloatTensor(final_dist).to(x.device)
+        final_labels = torch.LongTensor(final_labels).to(x.device)
+
+        # Start binary search
+        for outer_step in range(self.binary_search_steps):
+
+            self.global_step = 0
+
+            slack = nn.Parameter(x.clone())
+            newimg = x.clone()
```
Comment (gwding) [inline advertorch/attacks/ead.py]: `newimg` seems to be x^{(k)} in the paper? the name `newimg` is a bit misleading to me.
re: naming of variable. It seems that `y` is already used as label by default by the toolbox, so there's a name clash with the y in the paper.

how about instead of names like `newimg` `assign_newimg` `slack`, let's directly use the names from the paper, e.g. `xx_k xx_k_p_1 yy_k yy_k_p_1`

also for clarity, at the beginning of each iteration, we should let `yy_k = yy_k_p_1`

File: advertorch/attacks/ead.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+            optimizer = optim.SGD([slack], lr=self.learning_rate)
```
Comment (gwding) [inline advertorch/attacks/ead.py]: since the optimizer is simple gradient descent, maybe we should explicitly write it out? instead of using the `optim.SGD`?
In this case, the `polynomial_decay` logic can also be simplified.

**Comment by araujoalexandre** [comment]
Thank you for your review, I will modify the code in a few days !

File: advertorch/attacks/ead.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+        real = (y_onehot * output).sum(dim=1)
+        other = ((1.0 - y_onehot) * output - (y_onehot * TARGET_MULT)).max(1)[0]
+
+        if self.targeted:
+            loss1 = clamp(other - real + self.confidence, min=0.)
+        else:
+            loss1 = clamp(real - other + self.confidence, min=0.)
+
+        loss21 = l1dist.sum()
```
[the remark this answers is not in the document]
Comment (araujoalexandre) [reply advertorch/attacks/ead.py]: I rename the losses loss_logit loss_l1 loss_l2 as you proposed

File: advertorch/attacks/ead.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        loss2 = l2distsq.sum()
+
+        loss1 = torch.sum(const * loss1)
+
+        loss = loss1 + loss2 + (self.beta * loss21)
```
[the remark this answers is not in the document]
Comment (araujoalexandre) [reply advertorch/attacks/ead.py]: I merged the loss_opt_fn and the loss_fn in order to avoid duplication

File: advertorch/attacks/ead.py (hunk 1 above)
Comment (araujoalexandre) [reply advertorch/attacks/ead.py]: I rename the variable xx_k yy_k as you proposed, just to let you know, the slack and newimg variables name come from the original implementation of the authors : https://github.com/ysharma1126/EAD_Attack

File: advertorch/attacks/ead.py (hunk 2 above)
Comment (araujoalexandre) [reply advertorch/attacks/ead.py]: I removed the polynomial_decay function, i just update the learning rate after the gradient step.

File: advertorch/attacks/ead.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+        final_dist = torch.FloatTensor(final_dist).to(x.device)
+        final_labels = torch.LongTensor(final_labels).to(x.device)
+
+        # Start binary search
+        for outer_step in range(self.binary_search_steps):
+
+            self.global_step = 0
+
+            slack = nn.Parameter(x.clone())
```
[the remark this answers is not in the document]
Comment (araujoalexandre) [reply advertorch/attacks/ead.py]: I haven't managed to do that. The original implementation take the slack variable (i.e. y_k) as parameters to update. By taking y_k + delta with delta as parameters, the comparaison with cleverhans doesn't succeed.

[the remark this answers is not in the document]
Comment (gwding) [reply advertorch/attacks/ead.py]: I see. Maybe I missed something about the ISTA, or it needs special treatment. I'm ok with leave it like this.
